File: leetcode/2712.使所有字符相等的最小成本/Main.java
class Solution {
    int n;
    public long minimumCost(String s) {
        n = s.length();
        List<int[]> list = new ArrayList<>();
        char flag = s.charAt(0);
        int start = 0;
        for(int i=1;i<s.length();i++){
            char ch = s.charAt(i);
            if(ch!=flag){
                list.add(new int[]{start,i-1,flag-'0'});
                flag = ch;
                start=i;
            }
        }
        list.add(new int[]{start,s.length()-1,flag-'0'});
        if(list.size()==1)
            return 0L;
        return change(list);
    }

    public long change(List<int[]> list){
        long ans = 0;
        int left = 0;
        int right = list.size()-1;
        int flag1 = list.get(left)[2];
        int flag2 = list.get(right)[2];
        while (left<right){
            int[] temp1 = list.get(left);
            int[] temp2 = list.get(right);
            int val1 = 1+temp1[1];
            int val2 = this.n-temp2[0];
            if(left+1==right){
                if(flag1==flag2){
                    break;
                }
            }
            if(val1<=val2){
                left++;
                flag1 = list.get(left)[2];
                ans+=val1;
            }
            else{
                right--;
                flag2 = list.get(right)[2];
                ans+=val2;
            }
        }
        return ans;
    }
}

File: leetcode/2712.使所有字符相等的最小成本/main.cc
class Solution {
public:
    long long minimumCost(string s) {
        int n = s.length();
        vector<long long> dp(n+1,0);
        for(int i=2;i<=n;i++){
            char cur = s[i-1];
            char pre = s[i-2];
            if(cur == pre){
                dp[i] = dp[i-1];
            }else{
                dp[i] = dp[i-1] + min(i-1,n-(i-1));
            }
        }
        return dp[n];
    }
};